### PR TITLE
all: prep for path-based trie storage

### PR DESCRIPTION
**Pull request by holiman**
This PR moves some trie-related db accessor methods to a different file, and also removes the `schema` type. Instead of the schema type, a string is used to distinguish between hashbased/pathbased db accessors. 
This also moves some code from trie package to rawdb package. 

This PR is intended to be a no-functionality-change prep PR for #25963 .

### CODE REVIEW

File: eth/downloader/downloader.go [hunk 1]
```diff
@@ -229,7 +229,7 @@ func New(checkpoint uint64, stateDb ethdb.Database, mux *event.TypeMux, chain Bl
 		dropPeer:       dropPeer,
 		headerProcCh:   make(chan *headerTask, 1),
 		quitCh:         make(chan struct{}),
-		SnapSyncer:     snap.NewSyncer(stateDb, chain.TrieDB().Scheme()),
+		SnapSyncer:     snap.NewSyncer(stateDb, rawdb.HashScheme),
```
Comment (rjl493456442) [inline eth/downloader/downloader.go:232]: Let's use `chain.TrieDB().Scheme()`. The hardcoded string might be forgot to update in the future.

Comment (holiman) [reply eth/downloader/downloader.go:232]: Yep, this was a change that I made before I changed the `TrieDB().Scheme()` return, then forgot to undo it

Comment (holiman) [inline eth/downloader/downloader.go:232]:
```suggestion
		SnapSyncer:     snap.NewSyncer(stateDb, chain.TrieDB().Scheme()),
```

File: core/rawdb/schema.go [hunk 2]
```diff
@@ -100,12 +100,14 @@ var (
 	CodePrefix            = []byte("c") // CodePrefix + code hash -> account code
 	skeletonHeaderPrefix  = []byte("S") // skeletonHeaderPrefix + num (uint64 big endian) -> header
 
+	// Path-based trie node scheme.
+	trieNodeAccountPrefix = []byte("A") // trieNodeAccountPrefix + hexPath -> trie node
+	trieNodeStoragePrefix = []byte("O") // trieNodeStoragePrefix + accountHash + hexPath -> trie node
```
Comment (rjl493456442) [inline core/rawdb/schema.go:105]: These two are correspond to `SnapshotAccountPrefix` and  `SnapshotStoragePrefix `. And prefix `A` and `O` are not occupied by any other data, so they are picked.

**Review by rjl493456442 — APPROVED** [review]

File: core/rawdb/accessors_trie.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+// for archive node and some other tries(e.g. light trie).
+const HashScheme = "hashScheme"
+
+// PathScheme is the new path-based state scheme with which trie nodes are stored
+// in the disk with node path as the database key. This scheme will only store one
+// version of state data in the disk, which means that the state pruning operation
+// is native. At the same time, this scheme will put adjacent trie nodes in the same
+// area of the disk with good data locality property. But this scheme needs to rely
+// on extra state diffs to survive deep reorg.
+const PathScheme = "pathScheme"
```
Comment (karalabe) [inline core/rawdb/accessors_trie.go:170]: Please move constants up to the head of the file. It's very weird to have them deep down mixed with other code.

File: core/rawdb/accessors_trie.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+// HasTrieNode checks the trie node presence with the provided node info and
+// the associated node hash.
```
Comment (karalabe) [inline core/rawdb/accessors_trie.go:173]: Nitpick here and the rest of the generic methods below: "with the provided node info" is ambiguous given that based on `scheme` one or another set of parameters need to be provided. E.g. it's a bit strange that hash is always needed, even in path based storage. Perhaps we could more explicitly lay out what needs to be given and when?

**Comment by holiman** [comment]
> Do we drop the concept altogether long term, or will we need it forever (e.g. light clients)? If we will drop long term, string is fine. If we'll have to tiptoe around it forever, maaaaybe a typed constant would be better?

This is @rjl493456442 's code -- but my understanding is that we'll drop the hashScheme in a not too distant future. Doesn't make sense to keep them both once we're convinced that pathbased works reliably

File: core/rawdb/accessors_trie.go (hunk 4 above)
Comment (holiman) [reply core/rawdb/accessors_trie.go:173]: Fixed

**Comment by rjl493456442** [comment]
@karalabe I hope we can drop hash scheme. Hash scheme is still used by (1) les (2) archive mode. We can iterate LES protocol a bit to return the path along with the nodes, for archive node, I have an alternative approach to design a lighter archive mode based on the state history(flat data), but note, the lighter archive can't provide the merkle proof since trie nodes are gone.

Otherwise we can nuke out the hash scheme totally, and I don't think we have a big blocker of it.

**Pull request merged**